### PR TITLE
[query] improve type hinting for tables

### PR DESCRIPTION
**Pull request by danking**
Without these changes, my IDE cannot deduce that the return value of, say, `select` is a `Table` which prevents jump-to-definition on subsequent method calls.

I also received an error about `Table.collect` returning an `ArrayExpression` which it will only do when `_localize=False`. `overload` teaches this fact to the type system.

I didn't use the specific expression in `base_expression.py` because that would require an import circularity.

### CODE REVIEW

File: hail/python/hail/ir/ir.py [hunk 1]
```diff
@@ -2972,7 +2973,10 @@ def register_seeded_function(name, param_types, ret_type):
     _register(_seeded_function_registry, name, (param_types, ret_type))
 
 
-def udf(*param_types):
+T = TypeVar('T')
+
+
+def udf(*param_types) -> Callable[[Callable[..., T]], Callable[..., T]]:
```
Comment (patrick-schultz) [inline hail/python/hail/ir/ir.py:2979]: On `_make_dec` you added annotations on `wrapper`, but not here. Are they needed?

**Review by patrick-schultz — CHANGES_REQUESTED** [review]

File: hail/python/hail/expr/expressions/base_expression.py [hunk 2]
```diff
@@ -1101,6 +1101,12 @@ def take(self, n, _localize=True):
             return hl.eval(e)
         return e
 
+    @overload
+    def collect(self) -> List[Any]:
+        ...
+    @overload
+    def collect(self, _localize=False) -> 'Expression':
```
[the remark this answers is not in the document]
Comment (danking) [reply hail/python/hail/expr/expressions/base_expression.py:1108]: It introduces a circularity between base_expression and typed_expression, but mypy and Python appear OK with that, so I did it.

[the remark this answers is not in the document]
Comment (danking) [reply hail/python/hail/expr/expressions/base_expression.py:1108]: Ah, mypy is happy, but Python won't import hail now:
```ipython3

----> 1 import hail as hl

File ~/projects/hail/hail/python/hail/__init__.py:33
     14 __doc__ = r"""
     15     __  __     <>__
     16    / /_/ /__  __/ /
   (...)
     27 To report a bug, please open an issue: https://github.com/hail-is/hail/issues
     28 """
     30 # F403 'from .expr import *' used; unable to detect undefined names
     31 # F401 '.expr.*' imported but unused
     32 # E402 module level import not at top of file
---> 33 from .expr import *  # noqa: E402, F403
     34 from .expr import aggregators  # noqa: E402
     35 from hail.utils import (Struct, Interval, hadoop_copy, hadoop_open, hadoop_ls,  # noqa: E402
     36                         hadoop_stat, hadoop_exists, hadoop_is_file,
     37                         hadoop_is_dir, hadoop_scheme_supported, copy_log, ANY_REGION)

File ~/projects/hail/hail/python/hail/expr/__init__.py:8
      6 from .matrix_type import tmatrix
      7 from .blockmatrix_type import tblockmatrix
----> 8 from .expressions import (analyze, eval, eval_typed, eval_timed, extract_refs_by_indices, get_refs,
      9                           matrix_table_source, table_source, check_entry_indexed, check_row_indexed, Indices, Aggregation,
     10                           apply_expr, construct_expr, construct_variable, construct_reference, impute_type, to_expr,
     11                           cast_expr, unify_all, unify_types_limited, unify_types, unify_exprs, Expression,
     12                           ExpressionException, ArrayExpression, ArrayNumericExpression, BooleanExpression, CallExpression,
     13                           CollectionExpression, DictExpression, IntervalExpression, LocusExpression, NumericExpression,
     14                           Int32Expression, Int64Expression, Float32Expression, Float64Expression, SetExpression,
     15                           StreamExpression, StringExpression, StructExpression, TupleExpression, NDArrayExpression,
     16                           NDArrayNumericExpression, expr_any, expr_int32, expr_int64, expr_float32, expr_float64,
     17                           expr_call, expr_bool, expr_str, expr_locus, expr_interval, expr_array, expr_ndarray, expr_set,
     18                           expr_dict, expr_tuple, expr_struct, expr_oneof, expr_numeric, coercer_from_dtype)
     19 from .functions import (literal, chi_squared_test, if_else, cond, switch, case, bind, rbind,
     20                         contingency_table_test, dbeta, dict, dpois, exp, entropy, fisher_exact_test, gp_dosage,
     21                         hardy_weinberg_test, parse_locus, parse_variant, variant_str, locus, locus_from_global_position,
   (...)
     37                         _showstr, _sort_by, _compare, _locus_windows_per_contig, shuffle, _console_log, dnorm, dchisq,
     38                         query_table, keyed_union, keyed_intersection, repeat, _zip_join_producers)
     40 __all__ = ['HailType',
     41            'hail_type',
     42            'is_container',
   (...)
    313            'repeat',
    314            ]

File ~/projects/hail/hail/python/hail/expr/expressions/__init__.py:2
      1 from .indices import Indices, Aggregation
----> 2 from .base_expression import (ExpressionException, Expression, impute_type,
      3                               to_expr, cast_expr, unify_all, unify_types_limited, unify_types,
      4                               unify_exprs)
      5 from .typed_expressions import (ArrayExpression, ArrayStructExpression,
      6                                 ArrayNumericExpression, BooleanExpression, CallExpression,
      7                                 CollectionExpression, DictExpression, IntervalExpression, LocusExpression,
   (...)
     11                                 SetStructExpression,
     12                                 apply_expr, construct_expr, construct_variable, construct_reference)
     13 from .expression_typecheck import (expr_any, expr_int32, expr_int64,
     14                                    expr_float32, expr_float64, expr_call, expr_bool, expr_str, expr_locus,
     15                                    expr_interval, expr_array, expr_stream, expr_ndarray, expr_set, expr_dict, expr_tuple,
     16                                    expr_struct, expr_oneof, expr_numeric, coercer_from_dtype)

File ~/projects/hail/hail/python/hail/expr/expressions/base_expression.py:17
     15 from hail.utils.linkedlist import LinkedList
     16 from .indices import Indices, Aggregation
---> 17 from .tyepd_expressions import ArrayExpression
     19 from hail.expr.types import summary_type
     22 class Summary(object):

ModuleNotFoundError: No module named 'hail.expr.expressions.tyepd_expressions'

```

File: hail/python/hail/ir/ir.py (hunk 1 above)
Comment (danking) [reply hail/python/hail/ir/ir.py]: Yeah, OK, so, mypy lets things slide if it finds `Any` and decorator is being a bad actor here:
```
def decorator(
    caller: Callable[..., Any], _func: Callable[..., Any] | None = ...
) -> Callable[[Callable[..., Any]], Callable[..., Any]]: ...
```

I added hail_decorator.py which applies an actually useful type to `@decorator`. I then used it here. In the end, I'm not sure how much it buys us, but all this nonsense does mean that mypy knows that `_is_snp_transition` returns `BooleanExpression`.

**Review by patrick-schultz — APPROVED** [review]

**Pull request merged**